controllers/fleetnotification/fleetnotification_controller.go

```diff
@@ -55,21 +58,38 @@ func (r *ManagedFleetNotificationReconciler) Reconcile(ctx context.Context, requ
 	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
 	reqLogger.Info("Reconciling FleetNotification")
 
-	// TODO(user): your logic here
 	nr := ocmagentv1alpha1.ManagedFleetNotificationRecord{}
 
 	err := r.Client.Get(ctx, request.NamespacedName, &nr)
 	if err != nil {
 		return reconcile.Result{}, err
 	}
 
+	for n, rn := range nr.Status.NotificationRecordByName {
+		resendWait := rn.ResendWait
+		for i, ri := range rn.NotificationRecordItems {
+			// Consider the record is stale if the lastSendTime is older than resendWait + 15 days
+			eol := ri.LastTransitionTime.Time.Add(time.Duration(resendWait+NotificationRecordStaleTimeoutInHour) * time.Hour)
+			if time.Now().After(eol) {
+				log.Info(fmt.Sprintf("NotificationRecord for notification %s and hostedcluster %s has not been updated "+
+					"for a while and considered as stale, cleaning up...", rn.NotificationName, ri.HostedClusterID))
+
+				patch := []byte(fmt.Sprintf(`[{"op": "remove", "path": "/status/notificationRecordByName/%d/notificationRecordItems/%d"}]`, n, i))
```

Comment:
This is something where having a test written as part of the PR will really help to know for sure, but when I read this code it throws up a few red flags for me - if this patch succeeds and removes an item from the slice, then the loop index is not going to align with the reality of the object. I feel like it might end up removing the wrong items if there are multiple items for a notification record needing pruning.

Let me illustrate it with an example:

Let's say our CR looks like this:
/status/notificationRecordByName/0/notificationRecordItems/0  <--- will get removed 
/status/notificationRecordByName/0/notificationRecordItems/1  <--- will get removed
/status/notificationRecordByName/0/notificationRecordItems/2

On the first iteration of the inner loop, it does a patch to remove list item 0 (`i` will equal 0).

Now the CR looks like this:
/status/notificationRecordByName/0/notificationRecordItems/0  <--- will get removed
/status/notificationRecordByName/0/notificationRecordItems/1
but note that the in-memory CR that the loop is iterating through has not changed at all after this.

The next time it loops, it's going to want to delete the next entry in the list, but `i` is going to be 1, so it's going to remove:
/status/notificationRecordByName/0/notificationRecordItems/1

I could of course be completely off-base and wrong here.

I'm actually not sure if a test would really catch this, because the test is just mocking the patch and not really doing it..

Reply:
I gave this a quick test on a real cluster today, and yeah, it behaves the way I was expecting. Will have a think about a suggestion for a fix.

Reply:
hmm, thanks for the testing, that is not good...

Reply:
it should `return` after the status patch, not sure why it will continue in the loop.
How about if we add a `break` after the `r.Client.Status().Patch()`?

Reply:
Really I think the simplest solution might just be to end the reconcile after a single successful patch. So just a `return ctrl.Result{}, nil` right after the testing of the error condition.

The patch is going to force another reconcile immediately anyway, at which point it'll re-read the object in its new state.

Sure, there's a downside that it might end up reconciling a number of times in quick succession, but I don't think that's any major cause for concern.

This seemed to work fine when I tested it.

Reply:
> it should return after the status patch, not sure why it will continue in the loop.

I don't see that the code is currently doing that right now, but if you make it do that (as suggestd in my last reply) it will fix it.

Reply:
ah, I think I was wrong, it will return only when the `err != nil`, sorry for my brain is not fully functional yet.
So if as you tested that add another return on a succeeded patch operation, that would be the most easiest way for us.

Reply:
added a new return on the succeeded call

Reply:
Perfect, thanks! :+1: